src/namespaced_hash.rs

```diff
@@ -65,203 +64,273 @@ impl MerkleHash for NamespacedSha2Hasher {
+            min_ns: NamespaceId::default(),
+            max_ns: NamespaceId::default(),
+            hash: [0u8; HASH_LEN],
+        }
     }
 }
 
-impl NamespacedHash {
-    pub fn with_min_and_max_ns(min_namespace: NamespaceId, max_namespace: NamespaceId) -> Self {
-        let mut out = Self([0u8; NAMESPACED_HASH_LEN]);
-        out.0[0..NAMESPACE_ID_LEN].copy_from_slice(min_namespace.as_ref());
-        out.0[NAMESPACE_ID_LEN..2 * NAMESPACE_ID_LEN].copy_from_slice(max_namespace.as_ref());
-        out
+impl<const NS_ID_SIZE: usize> NamespacedHash<NS_ID_SIZE> {
+    pub const fn size() -> usize {
+        2 * NS_ID_SIZE + HASH_LEN
     }
-    pub fn min_namespace(&self) -> NamespaceId {
-        let mut out = [0u8; NAMESPACE_ID_LEN];
-        out.copy_from_slice(&self.0[..NAMESPACE_ID_LEN]);
-        NamespaceId(out)
+
+    pub const fn empty() -> Self {
+        Self {
+            min_ns: NamespaceId([0; NS_ID_SIZE]),
+            max_ns: NamespaceId([0; NS_ID_SIZE]),
+            hash: [
+                227, 176, 196, 66, 152, 252, 28, 20, 154, 251, 244, 200, 153, 111, 185, 36, 39,
+                174, 65, 228, 100, 155, 147, 76, 164, 149, 153, 27, 120, 82, 184, 85,
+            ],
+        }
     }
 
-    pub fn max_namespace(&self) -> NamespaceId {
-        let mut out = [0u8; NAMESPACE_ID_LEN];
-        out.copy_from_slice(&self.0[NAMESPACE_ID_LEN..2 * NAMESPACE_ID_LEN]);
-        NamespaceId(out)
+    pub fn with_min_and_max_ns(
+        min_ns: NamespaceId<NS_ID_SIZE>,
+        max_ns: NamespaceId<NS_ID_SIZE>,
+    ) -> Self {
+        Self {
+            min_ns,
+            max_ns,
+            ..Default::default()
+        }
     }
 
-    fn set_hash(&mut self, hash: &[u8]) {
-        self.0[2 * NAMESPACE_ID_LEN..].copy_from_slice(hash)
+    pub fn min_namespace(&self) -> NamespaceId<NS_ID_SIZE> {
+        self.min_ns
     }
 
-    pub fn empty() -> Self {
-        EMPTY_ROOT.clone()
+    pub fn max_namespace(&self) -> NamespaceId<NS_ID_SIZE> {
+        self.max_ns
     }
 
-    pub fn contains(&self, namespace: NamespaceId) -> bool {
+    fn set_hash(&mut self, new_hash: &[u8]) {
+        self.hash.copy_from_slice(new_hash)
+    }
+
+    pub fn contains(&self, namespace: NamespaceId<NS_ID_SIZE>) -> bool {
         self.min_namespace() <= namespace
             && self.max_namespace() >= namespace
             && !self.is_empty_root()
     }
 
     pub fn is_empty_root(&self) -> bool {
-        self == &EMPTY_ROOT
+        self == &Self::empty()
     }
 
-    pub fn hash_leaf(raw_data: impl AsRef<[u8]>, namespace: NamespaceId) -> Self {
+    pub fn hash_leaf(raw_data: impl AsRef<[u8]>, namespace: NamespaceId<NS_ID_SIZE>) -> Self {
         let mut output = NamespacedHash::with_min_and_max_ns(namespace, namespace);
         let mut hasher = Hasher::new_with_prefix(LEAF_DOMAIN_SEPARATOR);
-        hasher.update(namespace.as_ref());
+        hasher.update(namespace.as_ref()); // TODO: shouldn't this be hashed 2 times?
```

Comment:
I think, leaf here means Namespaced Data Item as shown [here](https://github.com/celestiaorg/nmt/blob/master/docs/spec/nmt.md). So only 1 namespace ID exists. You can remove the todo

Reply:
done